PythonCodeConventions.md

```diff
@@ -49,6 +58,45 @@ y = fib(2.5)
 In this case, mypy will inform that the call using the floating point value is not valid:
 __mypytest.py:10: error: Argument 1 to "fib" has incompatible type "float"; expected "int"__
 
+The following construct is _only_ allowed in case of circular dependencies, otherwise use regular import statements.
+``` python
+MYPY = False
+if MYPY:
+    from [namespace] import [class]
```

Comment:
I greatly prefer this solution, because it is less magic:

    from typing import TYPE_CHECKING
    if TYPE_CHECKING:
        from [package] import [class]

Reply:
It is suggested to drop the mechanism entirely.

Reply:
Not entirely. Currently it is suggested to drop the mechanism except in cases where you'll have a circular dependency. So if you drop it entirely then you'll either have circular dependencies and the code won't run, or you'll have to import it locally in a function which also violates the PEP8 rule to import everything at the top of the file.

Reply:
Yes, not entirely, bad choice of words. But in case of _absolutely unavoidable_ circular dependencies. Currently the Controller class is passed around everywhere, which isn't necessary in quite a lot of occurrences.  Using the above construct hides those dependencies and stand in the way of approving on this.

Reply:
Why would we use those only in case of circular dependencies? The code doesn't actually *need* to import it. It just needs it for the hinting. I'd actually argue that if the code doesn't need it, you should use the "softer" typing.

Reply:
Why does the distinction matter? Type hints are code too.

Reply:
I agree with @sedwards2009 to some degree.  At least in that imports necessary for typing are grouped with the 'regular' imports by default.

Reply:
In either case, regardless of when it is used, I would still prefer to use the `TYPE_CHECKING` variable rather than the magic `MYPY` variable.

Reply:
If it doesn't matter which one, why define that it should only be used for circular imports? Eg; Why should we *only* use them to prevent circular imports (and thus not for the case that I'm using them for?).
I have a slight preference to just leave the circular import statement out and only state that we should use the TYPE_CHECKING in case you only want to import for use with type checking (and don't enforce a reason why someone would want to do this)

Reply:
@nallath @Ghostkeeper I do agree that using the TYPE_CHECKING variant is the better option. I do not agree with allowing this for every import related to a type added just for type checking.

Reply:
The Application team discussed this before and we decided that in our team we don't want to use the MYPY / TYPE_CHECKING imports, unless you have to.

Reasoning: We had code where the conditional import made sense, but later code changed and did use the library instead of only for type checking. Normally PyCharm warns you for this error, but that doesn't work when this trick is used. These bugs are only detected run time.

Advantage of using the TYPE_CHECKING construct is minimal and to my opinion doesn't outweigh the hard bugs it can hide.

Reply:
Saying that you never will have circular dependencies is IMHO crazy. While should be avoided, there are cases where it makes sense, especially for type checking.

The TYPE_CHECKING flag was added later in the typing module, but it is already available in the older version we use in the printer firmware. Also, it's introduced to improve typechecking compatibility with different tools, instead of having a tool specific flag. So I would prefer to use that flag instead of the custom MYPY which locks to a specific tool.

Reply:
It seems the concensus is to prefer TYPE_CHECKING over MYPY? If so, let's settle and focus on other issues.

Reply:
@ChrisTerBeke Seems so, resolving conversation